lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+		blockNode.Status |= StatusBlockStored
+	}
+	// If the DB update fails, then we should return an error.
+	if err = bc.upsertBlockAndBlockNodeToDB(block, blockNode, false); err != nil {
+		return nil, errors.Wrapf(err, "storeValidateFailedBlockInBlockIndex: Problem upserting block and block node to DB")
+	}
+	return blockNode, nil
+}
+
+// storeCommittedBlockInBlockIndex upserts the blocks into the in-memory block index and updates its status to
+// StatusBlockCommitted. If the BlockNode does not have StatusBlockValidated and StatusBlockStored statuses,
+// we also add those. It also writes the block to the block index in badger by calling upsertBlockAndBlockNodeToDB.
+func (bc *Blockchain) storeCommittedBlockInBlockIndex(block *MsgDeSoBlock) (*BlockNode, error) {
+	blockNode, err := bc.getOrCreateBlockNodeFromBlockIndex(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "storeCommittedBlockInBlockIndex: Problem getting or creating block node")
+	}
+	// If the block is committed, then this is a no-op.
+	if blockNode.IsCommitted() {
+		return blockNode, nil
+	}
+	blockNode.Status |= StatusBlockCommitted
+	if !blockNode.IsValidated() {
+		blockNode.Status |= StatusBlockValidated
+	}
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	// If the DB update fails, then we should return an error.
+	if err = bc.upsertBlockAndBlockNodeToDB(block, blockNode, true); err != nil {
+		return nil, errors.Wrapf(err, "storeCommittedBlockInBlockIndex: Problem upserting block and block node to DB")
+	}
+	return blockNode, nil
+}
+
+// upsertBlockAndBlockNodeToDB writes the BlockNode to the blockIndexByHash in badger and writes the full block
+// to the db under the <blockHash> -> <serialized block> index.
+func (bc *Blockchain) upsertBlockAndBlockNodeToDB(block *MsgDeSoBlock, blockNode *BlockNode, storeFullBlock bool) error {
+	// Store the block in badger
+	err := bc.db.Update(func(txn *badger.Txn) error {
+		if bc.snapshot != nil {
+			bc.snapshot.PrepareAncestralRecordsFlush()
+			defer bc.snapshot.StartAncestralRecordsFlush(true)
+			glog.V(2).Infof("upsertBlockAndBlockNodeToDB: Preparing snapshot flush")
+		}
+		// TODO: Do we want to write the full block once.
+		// Store the new block in the db under the
+		//   <blockHash> -> <serialized block>
+		// index.
+		// TODO: In the archival mode, we'll be setting ancestral entries for the block reward. Note that it is
+		// 	set in PutBlockWithTxn. Block rewards are part of the state, and they should be identical to the ones
+		// 	we've fetched during Hypersync. Is there an edge-case where for some reason they're not identical? Or
+		// 	somehow ancestral records get corrupted?
+		if storeFullBlock {
+			if innerErr := PutBlockWithTxn(txn, bc.snapshot, block); innerErr != nil {
+				return errors.Wrapf(innerErr, "upsertBlockAndBlockNodeToDB: Problem calling PutBlock")
+			}
+		}
+
+		// TODO: if storeFullBlock = false, then we should probably remove the block from the DB? This can
```

Comment:
I don't think we need to worry too much about this case. It should be extremely rare for a block to pass our store-check without eventually being valid. Specifically, I think it'll be really hard for a block to get a valid QC without being actually valid.